### PR TITLE
Added Audio key

### PR DESCRIPTION
**Pull request by cberescu**
Using the OfflineAudioContext API created a key for the fingerprint. The mobile support for the OfflineAudioContext is not great so it jumps over if mobile is detected.

Tested also the AudioContext API but does not work all the time. 

Code and info about Audio Fingerprint found here: https://audiofingerprint.openwpm.com/

### CODE REVIEW

**Comment by Valve** [comment]
Oh, thanks, I saw this demo page, very cool. However I emailed the author to ask for a permission to integrate his code into FPJS2, because I wasn't sure about the licensing, but he didn't respond.
Do you know anything about if we can use his code?

**Comment by cberescu** [comment]
Well, it says : 

> This page is a part of Princeton CITP's Web Transparency and Accountability Project.  [https://webtap.princeton.edu/](url)

On the bottom of the Project page you can find the license link [http://creativecommons.org/licenses/by/3.0/us/](url)

So, i think that license applies to this also but i'm not 100% sure.

File: fingerprint2.js [hunk 1]
```diff
@@ -105,17 +105,19 @@
...
-        that.each(newKeys, function(pair) {
-          var value = pair.value;
-          if (typeof pair.value.join !== "undefined") {
-            value = pair.value.join(";");
-          }
-          values.push(value);
-        });
-        var murmur = that.x64hash128(values.join("~~~"), 31);
-        return done(murmur, newKeys);
+      this.audioKey(keys, function(newKeys){
```
Comment (Valve) [inline fingerprint2.js]: Seems like you didn't run Gulp to verify spaces vs tabs etc

Comment (cberescu) [inline fingerprint2.js]: Yeap, i did it. No erros found : 
https://www.dropbox.com/s/dmjym44qr0eeaop/Screenshot%202016-05-23%2022.06.22.png?dl=0

**Comment by Valve** [comment]
Thanks, should be a Github formatting issues

**Comment by m1johnson** [comment]
Nice find! This looks very interesting.

**Comment by englehardt** [comment]
Me and @dreisman are the authors of the scripts on https://audiofingerprint.openwpm.com/. We're fine with you using any of our code, but we'd like to note that several of the techniques are adapted directly from scripts found during our study, allowing us to examine the effectiveness of their implementations. These scripts are listed in the associated [technical report](http://randomwalker.info/publications/OpenWPM_1_million_site_tracking_measurement.pdf) or in the comments of the test page script.

I'd also like to note that we haven't yet verified the stability or uniqueness of the fingerprint. We'll do so in a follow up study.

**Comment by Valve** [comment]
Thanks!

File: fingerprint2.js [hunk 2]
```diff
@@ -498,6 +500,56 @@
         done(keys);
       }, 1);
     },
+	audioKey: function(keys, done){
+		if ((!this.options.excludeAudio) && !(("ontouchstart" in window) || (navigator.maxTouchPoints > 0) || (navigator.msMaxTouchPoints > 0))) {
```
Comment (Valve) [inline fingerprint2.js]: Why are you checking for touch events and touch points?

Comment (cberescu) [inline fingerprint2.js]: Sorry forthe late reply,

Checking if it is a mobile/tablet, on my tests the audio fingerprint didn't work well on them.

**Comment by drstyv** [comment]
Any new updates on this new feature yet?

**Comment by Valve** [comment]
Sorry, no updates. This feature is not a simple "add and forget" it requires testing in numerous browsers and on numerous platforms. Unluckily, I currently don't have enough free time to dedicate to this.

File: fingerprint2.js (hunk 2 above)
Comment (cachitocastania) [inline fingerprint2.js]: How would you rate the entropy of the plugin adding this audio fg? Would it be unique??

**Comment by Valve** [comment]
closing for now as a non-feature. The goal now is to stabilize what we have, not to add new FP sources which may destabilize the generation process. We may return to it in the future though.

**Pull request closed without merging**